File: packages/acp/src/fake-agent.ts
import type { AnyMessage, AnyRequest } from "@agentclientprotocol/sdk"
import { isJsonRpcRequest } from "./jsonrpc-line.js"
import { JSON_RPC_METHOD_NOT_FOUND } from "./methods.js"
import { encodeNdjsonChunk } from "./ndjson.js"

export type FakeAgentOptions = {
  readonly sessionId?: string
  readonly lockLoad?: boolean
  readonly pairSetModeWithTerminal?: boolean
}

export class FakeGrokAgent {
  readonly sessionId: string
  readonly lockLoad: boolean
  readonly pairSetModeWithTerminal: boolean
  readonly updates: Array<AnyMessage> = []

  constructor(options: FakeAgentOptions = {}) {
    this.sessionId = options.sessionId ?? "sess_test"
    this.lockLoad = options.lockLoad ?? false
    this.pairSetModeWithTerminal = options.pairSetModeWithTerminal ?? false
  }

  repliesFor(message: AnyMessage): ReadonlyArray<AnyMessage> {
    if (!isJsonRpcRequest(message)) return []
    return this.repliesForRequest(message)
  }

  encodeReplies(message: AnyMessage): Uint8Array {
    const replies = this.repliesFor(message)
    if (replies.length === 0) return new Uint8Array()
    return encodeNdjsonChunk(replies)
  }

  private repliesForRequest(request: AnyRequest): ReadonlyArray<AnyMessage> {
    switch (request.method) {
      case "initialize":
        return [ok(request.id, {
          protocolVersion: 1,
          agentCapabilities: { loadSession: true },
        })]
      case "session/new":
        return [
          notify("session/update", {
            sessionId: this.sessionId,
            update: {
              sessionUpdate: "available_commands_update",
              availableCommands: [
                { name: "compact", description: "Compact context" },
                { name: "always-approve", description: "Skip permission prompts" },
              ],
            },
          }),
          ok(request.id, {
            sessionId: this.sessionId,
            modes: {
              currentModeId: "normal",
              availableModes: [
                { id: "normal", name: "Normal" },
                { id: "plan", name: "Plan" },
                { id: "always-approve", name: "Always-approve" },
              ],
            },
          }),
        ]
      case "session/load":
        if (this.lockLoad) {
          return [fail(request.id, JSON_RPC_METHOD_NOT_FOUND, "session locked")]
        }
        return [ok(request.id, { sessionId: this.sessionId })]
      case "session/set_mode": {
        const result = ok(request.id, {})
        if (!this.pairSetModeWithTerminal) return [result]
        return [
          result,
          requestMsg("terminal/create", {
            sessionId: this.sessionId,
            command: "rm",
            args: ["-rf", "/tmp/beard-probe"],
          }, "term-1"),
        ]
      }
      case "session/prompt":
        return [
          notify("session/update", {
            sessionId: this.sessionId,
            update: {
              sessionUpdate: "agent_thought_chunk",
              content: { type: "text", text: "Considering the selection.\n" },
            },
          }),
          notify("session/update", {
            sessionId: this.sessionId,
            update: {
              sessionUpdate: "agent_thought_chunk",
              content: { type: "text", text: "Then I'll answer.\n" },
            },
          }),
          notify("session/update", {
            sessionId: this.sessionId,
            update: {
              sessionUpdate: "agent_message_chunk",
              content: { type: "text", text: "hello" },
            },
          }),
          notify("session/update", {
            sessionId: this.sessionId,
            update: {
              sessionUpdate: "tool_call",
              toolCallId: "call_1",
              title: "Edit",
              kind: "edit",
              status: "pending",
              content: [{
                type: "diff",
                path: "/tmp/file.ts",
                oldText: "old",
                newText: "new",
              }],
            },
          }),
          ok(request.id, { stopReason: "end_turn" }),
        ]
      case "session/request_permission":
        return [ok(request.id, { outcome: { outcome: "selected", optionId: "allow-once" } })]
      default:
        return [fail(request.id, JSON_RPC_METHOD_NOT_FOUND, `Method not found: ${request.method}`)]
    }
  }
}

export const requestPermissionEdit = (
  sessionId: string,
  id: string | number = "perm-1",
): AnyMessage =>
  requestMsg("session/request_permission", {
    sessionId,
    toolCall: {
      toolCallId: "call_1",
      title: "Edit",
      kind: "edit",
      content: [{ type: "diff", path: "/tmp/file.ts", oldText: "old", newText: "new" }],
    },
    options: [
      { optionId: "allow-once", name: "Allow once", kind: "allow_once" },
      { optionId: "reject-once", name: "Reject", kind: "reject_once" },
    ],
  }, id)

const ok = (id: string | number | null, result: unknown): AnyMessage => ({
  jsonrpc: "2.0",
  id,
  result,
})

const fail = (id: string | number | null, code: number, message: string): AnyMessage => ({
  jsonrpc: "2.0",
  id,
  error: { code, message },
})

const notify = (method: string, params: unknown): AnyMessage => ({
  jsonrpc: "2.0",
  method,
  params,
})

const requestMsg = (method: string, params: unknown, id: string | number): AnyMessage => ({
  jsonrpc: "2.0",
  id,
  method,
  params,
})

File: beard/core/src/main/scala/groksbeard/core/FakeAgent.scala
package groksbeard.core

final class FakeAgent(
    val sessionId: String = "sess_test",
    pairSetModeWithTerminal: Boolean = false,
    lockLoad: Boolean = false,
):
  def replies(msg: Rpc): List[Rpc] =
    msg match
      case Rpc.Request(id, method, _) => repliesFor(id, method)
      case _                          => Nil

  def encodeReplies(msg: Rpc): String =
    Ndjson.encodeChunk(replies(msg).map(Rpc.toLine))

  private def repliesFor(id: RpcId, method: String): List[Rpc] =
    method match
      case "initialize" =>
        List(Rpc.ok(id, InitializeResult(1, AgentCapabilities(loadSession = true)).asJson))
      case "session/new" =>
        List(
          Rpc.notifyOf(
            "session/update",
            AcpSessionNotify(
              sessionId,
              AcpUpdate.Commands(
                List(
                  SlashCommand("compact", "Compact context"),
                  SlashCommand("always-approve", "Skip permission prompts"),
                )
              ),
            ),
          ),
          Rpc.ok(
            id,
            SessionNewResult(
              sessionId,
              Some(
                SessionModeState(
                  "normal",
                  List(
                    ModeOption("normal", "Normal"),
                    ModeOption("plan", "Plan"),
                    ModeOption("auto", "Auto"),
                    ModeOption("always-approve", "Always approve"),
                  ),
                )
              ),
            ).asJson,
          ),
        )
      case "session/load" =>
        if lockLoad then List(Rpc.fail(id, Rpc.MethodNotFound, "session locked"))
        else List(Rpc.ok(id, SessionLoadResult(sessionId).asJson))
      case "session/set_mode" =>
        val result = Rpc.ok(id, EmptyObject().asJson)
        if !pairSetModeWithTerminal then List(result)
        else
          List(
            result,
            Rpc.request(
              RpcId.Str("term-1"),
              "terminal/create",
              TerminalCreateParams(sessionId, "rm", List("-rf", "/tmp/beard-probe")),
            ),
          )
        end if
      case "session/prompt" =>
        List(
          thought("Considering the selection.\n"),
          thought("Then I'll answer.\n"),
          agent("hello"),
          Rpc.notifyOf(
            "session/update",
            AcpSessionNotify(
              sessionId,
              AcpUpdate.ToolCall(
                toolCallId = "call_1",
                title = "Edit Main.scala",
                kind = "edit",
                status = "pending",
                content = List(
                  AcpContent.Diff(
                    path = "/tmp/Main.scala",
                    oldText = Some("object Main"),
                    newText = Some("object Main:\n  def run = ()"),
                  )
                ),
              ),
            ),
          ),
          Rpc.ok(id, SessionPromptResult("end_turn").asJson),
        )
      case _ =>
        List(Rpc.fail(id, Rpc.MethodNotFound, s"Method not found: $method"))

  private def thought(text: String): Rpc.Notify =
    chunk(AcpUpdate.Thought(AcpContent.Text(text)))

  private def agent(text: String): Rpc.Notify =
    chunk(AcpUpdate.Agent(AcpContent.Text(text)))

  private def chunk(update: AcpUpdate): Rpc.Notify =
    Rpc.notifyOf("session/update", AcpSessionNotify(sessionId, update))
end FakeAgent
